File: mobile/src/core/export/gpxExporter.ts
import { File, Paths } from "expo-file-system";
import type { Trip } from "../storage/models";

export async function exportTripGPX(trip: Trip): Promise<string | undefined> {
  const coords = trip.routeData ?? [];
  let gpx = `<?xml version="1.0" encoding="UTF-8"?>\n<gpx version="1.1" creator="QuickCar" xmlns="http://www.topografix.com/GPX/1/1">\n<trk><name>${trip.id}</name><trkseg>\n`;
  coords.forEach((c) => {
    const t = new Date(c.t ?? trip.startedAt).toISOString();
    gpx += `<trkpt lat="${c.lat}" lon="${c.lon}"><time>${t}</time></trkpt>\n`;
  });
  gpx += "</trkseg></trk></gpx>";
  try {
    const file = new File(Paths.cache, `trip-${trip.id}.gpx`);
    file.create();
    file.write(gpx);
    return file.uri;
  } catch {
    return undefined;
  }
}

File: QuickCar/Core/Export/GPXExporter.swift
import Foundation
import CoreLocation

enum GPXExporter {
    static func export(trip: Trip) -> URL? {
        let coords = RouteSimplifier.decode(trip.routeData)
        let iso = ISO8601DateFormatter()
        var gpx = """
        <?xml version="1.0" encoding="UTF-8"?>
        <gpx version="1.1" creator="QuickCar" xmlns="http://www.topografix.com/GPX/1/1">
        <trk><name>\(trip.id.uuidString)</name><trkseg>
        """
        for (index, c) in coords.enumerated() {
            let t = trip.startedAt.addingTimeInterval(Double(index))
            gpx += "<trkpt lat=\"\(c.latitude)\" lon=\"\(c.longitude)\"><time>\(iso.string(from: t))</time></trkpt>\n"
        }
        gpx += "</trkseg></trk></gpx>"
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("trip-\(trip.id.uuidString).gpx")
        do {
            try gpx.write(to: url, atomically: true, encoding: .utf8)
            return url
        } catch {
            Log.error("GPX export failed: \(error)")
            return nil
        }
    }
}
